### PR TITLE
Aggregations cleanup

### PR DESCRIPTION
**Pull request by angrykoala**
# Description

Unifies the behaviour between nested and top-level aggregations. This means that `reduce` is no longer used in top level, and subqueries are used instead

This causes a slight performance increase on string aggregations (longest, shortest) as we do not rely on `reduce` anymore but causes a slight performance decrease on top level aggregations when querying multiple fields, for example:

```graphql
query {
  moviesAggregate {
    title { longest }
    year { max } 
  }
}
```

OLD:
```
┌────────────────────────────────────────────────────────┬─────────┬────────┬─────────────────────────┬───────────┐
│                        (index)                         │ maxRows │ dbHits │          cache          │ time (ms) │
├────────────────────────────────────────────────────────┼─────────┼────────┼─────────────────────────┼───────────┤
│  aggregations.TopLevelAggregateWithMultipleFields_only │  1134   │  4537  │ { hits: 20, misses: 0 } │    21     │
```

NEW
```
┌───────────────────────────────────────────────────────────┬─────────┬────────┬─────────────────────────┬───────────┐
│                          (index)                          │ maxRows │ dbHits │          cache          │ time (ms) │
├───────────────────────────────────────────────────────────┼─────────┼────────┼─────────────────────────┼───────────┤
│   aggregations.TopLevelAggregateWithMultipleFields_only   │  1134   │  6802  │ { hits: 40, misses: 0 } │    28     │

```

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4285.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
| 🟦 aggregations.TopLevelAggregateWithMultipleFields | 6802 | N/A | 40 | N/A | 1134 |
| 🟦 aggregations.InterfacesAggregations | 6242 | N/A | 38 | N/A | 2080 |
| 🟦 aggregations.InterfacesAggregationsWithTwoFields | 11444 | N/A | 62 | N/A | 2080 |


🟥 - Performance worsened (dbHits)
🟩 - Performance improved (dbHits)
🟦 - New test
        

<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3403 | 32 | 36 | 1134 |
| 🟦 aggregations.TopLevelAggregateWithMultipleFields | 6802 | N/A | 40 | N/A | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 50 | 70 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 46 | 47 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 2282 | 2545 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2270 | 2308 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 29 | 45 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 30 | 48 | 2174 |
|  aggregations.DeeplyNestedCount | 11062443 | 11062443 | 2951 | 2824 | 2008534 |
| 🟦 aggregations.InterfacesAggregations | 6242 | N/A | 38 | N/A | 2080 |
| 🟦 aggregations.InterfacesAggregationsWithTwoFields | 11444 | N/A | 62 | N/A | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 137 | 140 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 89 | 69 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 143 | 190 | 3003 |
|  connections.Connection | 14082 | 14082 | 49 | 81 | 2174 |
|  connections.NestedConnection | 41118 | 41118 | 84 | 91 | 4516 |
|  create.SimpleMutation | 7 | 7 | 45 | 50 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 46 | 27 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 233 | 302 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 198 | 192 | 1183 |
|  2871.NestedRelationshipFilter | 23123 | 23170 | 66 | 63 | 4436 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 38 | 41 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 73 | 91 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 34 | 40 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 47 | 63 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 18 | 20 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 32 | 38 | 2174 |
|  query.QueryWhere | 9706 | 9723 | 28 | 35 | 2164 |
|  query.SimpleQueryWithNestedWhere | 9884 | 9901 | 41 | 52 | 2164 |
|  query.Nested | 10096041 | 10096041 | 5874 | 6308 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 5841 | 6224 | 2004000 |
|  query.OrFilterOnRelationships | 42006 | 42676 | 189 | 257 | 1635 |
|  query.OrFilterOnRelationshipsAndNested | 36360 | 35553 | 186 | 255 | 1635 |
|  query.QueryWithNestedIn | 13364 | 13458 | 48 | 53 | 1205 |
|  query.NestedConnectionWhere | 9834 | 9834 | 51 | 53 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9893 | 9849 | 78 | 79 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19152 | 19056 | 111 | 147 | 1598 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 148 | 182 | 1134 |
|  query.Fulltext | 80 | 80 | 28 | 34 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 35 | 45 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14092 | 14092 | 43 | 45 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14092 | 14092 | 52 | 61 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14190 | 14245 | 51 | 66 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14190 | 14245 | 75 | 81 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14629 | 14971 | 91 | 110 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14473 | 14670 | 90 | 97 | 2174 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 52 | 62 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14580 | 14490 | 69 | 82 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 39 | 45 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 67 | 84 | 4516 |
|  sorting.ConnectionWithSort | 3284 | 3284 | 56 | 63 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 57 | 63 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 51 | 55 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 299 | 244 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 217 | 240 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 86 | 104 | 2002 |

</details>
        
Old Schema Generation: 24.418s
Schema Generation: 24.741s
Old Subgraph Schema Generation: 27.780s
Subgraph Schema Generation: 28.141s

File: packages/graphql/src/translate/queryAST/ast/operations/AggregationOperation.ts [hunk 1]
```diff
@@ -96,186 +96,321 @@ export class AggregationOperation extends Operation {
...
+            extraSelectionWith,
+            sortClause,
+            ret
+        );
     }
 
-    public setEdgeFields(fields: AggregationField[]) {
-        this.edgeFields = fields;
-    }
+    // public transpile(context: QueryASTContext): OperationTranspileResult {
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/operations/AggregationOperation.ts:279]: Commented code

**Review by MacondoExpress — COMMENTED** [review]
Really love the output of this, let me have a better understanding of `context.direction` attribute but apart from that it looks good to me!

File: packages/graphql/src/translate/queryAST/ast/operations/AggregationOperation.ts [hunk 2]
```diff
@@ -96,186 +96,321 @@ export class AggregationOperation extends Operation {
...
-            extraSelection = [topLevelMatch, ...extraSelection];
+    private getPattern(context: QueryASTContext): Cypher.Pattern {
+        if (!context.target) {
+            throw new Error("Not Target");
         }
-
-        if (filterPredicates) {
-            clause.where(filterPredicates);
+        if (context.relationship) {
+            if (!context.direction || !context.source) {
```
[the remark this answers is not in the document]
Comment (angrykoala) [reply packages/graphql/src/translate/queryAST/ast/operations/AggregationOperation.ts:168]: Basically, the relationship that we have in the context is not really too useful without the direction, so I added those values into the context and this is just checking that those exist. We will want to improve this to make it more consistent

**Pull request merged**